File: R/LINTUL1.R
# Author: Robert J. Hijmans, r.hijmans@gmail.com
# License GPL3
# Version 0.1  July 2016



lintul1 <- function(crop, control, weather) {
	m <- Lintul1Model$new()
	if (!missing(crop)) { crop(m) <- crop }
	if (!missing(control)) { control(m) <- control }
	if (!missing(weather)) { weather(m) <- weather }
	return(m)
}


setMethod("run", signature('Rcpp_Lintul1Model'), 
	function(x, ...) {
		x$run()
		#ff <- names(getRefClass("Rcpp_Lintul1Output")$fields())
		out <- x$out
		date <- as.Date(x$control$emergence, origin="1970-01-01") + out$step
		v <- data.frame(date, out$DLV, out$LAI, out$TSUM, out$WRT, out$WLV, out$WLVD, out$WLVG, out$WST, out$WSO)	
		colnames(v) <- c("date", "DLV", "LAI", "TSUM", "WRT", "WLV", "WLVD", "WLVG", "WST", "WSO")
		v$Wtot <- v$WRT + v$WLVD + v$WLVG + v$WST + v$WSO
		v
	}
)


setMethod("crop<-", signature('Rcpp_Lintul1Model', 'list'), 
	function(x, value) {
		parameters <- c("LAIi", "SLA", "Tbase", "RGRL", "Tsum1", "Tsum2", "LAIcr", "RDRSHM", "RUE", "K", "RDRT", "FRTTB", "FLVTB", "FSTTB", "FSOTB")
		nms <- names(value)
		if (!all(parameters %in% nms)) stop(paste("parameters missing:", paste(parameters[!(parameters %in% nms)], collapse=", ")))
		value <- value[parameters]
		nms <- names(value)
		lapply(1:length(value), function(i) eval(parse(text = paste0("x$crop$", nms[i], " <- ", value[i]))))
		return(x)
	}
)


setMethod("weather<-", signature('Rcpp_Lintul1Model', 'list'), 
	function(x, value) {
		parameters <- c("date", "srad", "tmin", "tmax")
		nms <- names(value)
		if (!all(parameters %in% nms)) stop(paste("parameters missing:", paste(parameters[!(parameters %in% nms)], collapse=", ")))
		x$setWeather(value$date, value$tmin, value$tmax, value$srad, value$prec, value$wind, value$vapr)
		return(x)
	}
)


setMethod("weather<-", signature('Rcpp_Lintul1Model', 'list'), 
	function(x, value) {
		parameters <- c("date", "srad", "tmin", "tmax")
		nms <- names(value)
		if (!all(parameters %in% nms)) stop(paste("parameters missing:", paste(parameters[!(parameters %in% nms)], collapse=", ")))
		x$setWeather(value$date, value$tmin, value$tmax, value$srad)
		return(x)
	}
)


setMethod("control<-", signature('Rcpp_Lintul1Model', 'list'), 
	function(x, value) {
		#parameters <- c("emergence", "maxdur", "long_output")
		parameters <- c("emergence", "maxdur")
		nms <- names(value)
		if (!all(parameters %in% nms)) stop(paste("parameters missing:", paste(parameters[!(parameters %in% nms)], collapse=", ")))
		value <- value[parameters]
		nms <- names(value)
		lapply(1:length(value), function(i) eval(parse(text = paste0("x$control$", nms[i], " <- ", value[i]))))
		return(x)
	}
)


lintul1_crop <- function() {
	params <- list(LAIi=0.012, SLA=0.022, Tbase=0, RGRL=0.009, Tsum1=1110, Tsum2=2080, LAIcr=4, RDRSHM=0.03, RUE=3.0, K=0.6)
	RDRT <- matrix(c(-10.,0.03, 10.,0.03, 15.,0.04, 30.,0.09, 50.,0.09), ncol=2, byrow=TRUE)
	FRTTB <- matrix(c(0.,0.50, 110, 0.50, 275 ,0.34, 555, 0.12, 780, 0.07, 1055, 0.03, 1160, 0.02, 1305, 0, 2500, 0), ncol=2, byrow=TRUE)
	FLVTB <- matrix(c(0, 0.33, 110, 0.33, 275, 0.46, 555, 0.44, 780, 0.14, 1055, 0, 2500, 0), ncol=2, byrow=TRUE)
	FSTTB <- matrix(c(0, 0.17, 110, 0.17, 275, 0.20, 555, 0.44, 780, 0.79, 1055.,0.97, 1160, 0, 2500, 0), ncol=2, byrow=TRUE)
	FSOTB <- matrix(c(0, 0, 1055, 0, 1160.,0.98, 1305, 1, 2500, 1), ncol=2, byrow=TRUE)
	crop <- c(params, list(RDRT=RDRT, FRTTB=FRTTB, FLVTB=FLVTB, FSTTB=FSTTB, FSOTB=FSOTB))
	return(crop)
}


setMethod ('show' , 'Rcpp_Lintul1Model', function(object) { utils::str(object) } )	
setMethod ('show' , 'Rcpp_Lintul1Output', function(object) { utils::str(object) } )	
setMethod ('show' , 'Rcpp_Lintul1Crop', function(object) { utils::str(object) } )	
setMethod ('show' , 'Rcpp_Lintul1Control', function(object) { utils::str(object) })	
setMethod ('show' , 'Rcpp_Lintul1Weather', function(object) { utils::str(object) })	


readLIN1output <- function(f) {
	X <- readLines(f)
	H <- unlist(strsplit(X[7], '\t'))
	X <- X[-c(1:8)]
	X = .trim2(X)
	X <- strsplit(X, '\t')
	s = lapply(X, function(i) as.numeric(i))
	ss = do.call(rbind, s)
	colnames(ss) = H
	ss
}

File: src/LINTUL1.cpp
/*
Author Robert Hijmans
Date: May 2016
License: GNU General Public License (GNU GPL) v. 2 
*/

using namespace std;
#include <vector>
#include <algorithm>
#include <cmath>

#include "SimUtil.h"
#include "LINTUL1.h"

void Lintul1Model::output_initialize() {
	out.step.resize(0);
	out.TSUM.resize(0);
	out.DLV.resize(0);
	out.LAI.resize(0);
	out.WLVD.resize(0);
	out.WLV.resize(0);
	out.WLVG.resize(0);
	out.WST.resize(0);
	out.WRT.resize(0);
	out.WSO.resize(0);
}


void Lintul1Model::model_output(){
	out.step.push_back(step);
	out.TSUM.push_back(crop.TSUM);
	out.DLV.push_back(crop.DLV);
	out.LAI.push_back(crop.LAI);
	out.WLVD.push_back(crop.WLVD);
	out.WLV.push_back(crop.WLV);
	out.WLVG.push_back(crop.WLVG);
	out.WST.push_back(crop.WST);
	out.WRT.push_back(crop.WRT);
	out.WSO.push_back(crop.WSO);
}


void Lintul1Model::weather_step() {
	Tavg = (wth.tmin[time] + wth.tmax[time] ) / 2;
	Teff = std::max(0.0, Tavg - crop.Tbase);
}


void Lintul1Model::crop_initialize() {
	crop.TSUM = 0;
	crop.LAI = 0;
	crop.WLVD = 0;
	crop.WST = 0;
	crop.WSO = 0;
	crop.WLV = 0;
	crop.WRT = 0;
	crop.WLVi = crop.LAIi / crop.SLA;
	crop.WLVG = crop.WLVi;
	crop.alive = true;
}



void Lintul1Model::model_initialize() {
	step = 0;
	crop_initialize();
	output_initialize();
	crop.emergday = true;	
	time = control.emergence - wth.date[0];
	// need to check for out of bounds times (before of after start)
	// stop if time < 0 || time > wth.date.size();
}



void Lintul1Model::crop_rates() {
	  // rates  
	double RDR, RDRDV, RDRSH; // leaf relative death rate / from ageing / from self-shading
	double dLAI, gLAI; // LAI rates
	double PARint; // intercepted radiation
	double FRT, FLV, FST, FSO; // partitioning
	double GTotal;
  
    // radiation interception
	PARint = 0.5 * wth.srad[time] * (1 - exp(-crop.K * crop.LAI));
    
    // Total crop growth rate				
	GTotal = crop.RUE * PARint;
    
    // Development stage dependent partitioning of dry matter to organs
	FRT = approx(crop.FRTTB, crop.TSUM);
	FLV = approx(crop.FLVTB, crop.TSUM);
	FST = approx(crop.FSTTB, crop.TSUM);
	FSO = approx(crop.FSOTB, crop.TSUM);
    
    
    // LAI growth rate	
	if (crop.emergday) {
		gLAI = crop.LAIi;
		crop.emergday = false;
	} else if ((crop.TSUM < 330) & (crop.LAI < 0.75)) { 
		gLAI = crop.LAI * (exp(crop.RGRL * Teff) - 1);
	} else {
		gLAI = crop.SLA * crop.GLV;
	}
    
    // LAI senescence
    // relative death rate from development (ageing) 
	RDRDV = (crop.TSUM - crop.Tsum1) < 0 ? 0 : approx(crop.RDRT, Tavg);
    // relative death rate from self-shading
	RDRSH = clamp(0, crop.RDRSHM, crop.RDRSHM * (crop.LAI - crop.LAIcr) / crop.LAIcr);
    // relative death rate
	RDR = std::max(RDRDV, RDRSH);
    // loss of LAI				
	dLAI = crop.LAI * RDR;
    // rate of LAI change		
	crop.rLAI = gLAI - dLAI;
    // death rate (by weight)
	crop.DLV = crop.WLVG * RDR;
    
	crop.RWLVG = GTotal * FLV - crop.DLV;
	crop.RWST = GTotal * FST;
	crop.RWSO = GTotal * FSO;
	crop.RWRT = GTotal * FRT;
	crop.GLV = GTotal * FLV;
	
	if (crop.TSUM > crop.Tsum2) {
		crop.alive = false;
	} else if (wth.tmax[time] <  -10) {
		crop.alive = false;	
	}
	
}

	

void Lintul1Model::crop_states() {
  
	crop.LAI  = crop.LAI + crop.rLAI;
	crop.WLVG = crop.WLVG + crop.RWLVG;
	crop.WLVD = crop.WLVD + crop.DLV;
	crop.WLV  = crop.WLVG + crop.WLVD;
	crop.WST  = crop.WST + crop.RWST;
	crop.WSO  = crop.WSO + crop.RWSO;
	crop.WRT  = crop.WRT + crop.RWRT;

	crop.TSUM = crop.TSUM + Teff;  
}

	
	
void Lintul1Model::model_run() {
  
	model_initialize(); 
				
	while ((crop.alive) & (step < control.maxdur)) {  
		weather_step();
		crop_rates();
		model_output();
		crop_states();
		time++;
		step++;		
	}
}

